presto-main/src/main/java/io/prestosql/operator/scalar/ConcatWsFunction.java

```diff
@@ -0,0 +1,266 @@
+                    true,
+                    ImmutableList.of(new FunctionArgumentDefinition(false), new FunctionArgumentDefinition(false)),
+                    false,
+                    true,
+                    "Concatenates array elements using separator",
+                    SCALAR));
+        }
+
+        @Override
+        public ScalarFunctionImplementation specialize(FunctionBinding functionBinding)
+        {
+            if (functionBinding.getArity() != 2) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "There must be two arguments");
+            }
+
+            Type arrayValueType = functionBinding.getTypeVariable("T");
+            if (!(arrayValueType instanceof VarcharType) && !(arrayValueType instanceof UnknownType)) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Array values must be varchar types");
+            }
+
+            return new ScalarFunctionImplementation(
+                    NULLABLE_RETURN,
+                    ImmutableList.of(NEVER_NULL, NEVER_NULL),
+                    methodHandle(ConcatWsFunction.class, "concatWsArray", Slice.class, Block.class),
+                    Optional.empty());
+        }
+    }
+
+    public ConcatWsFunction()
+    {
+        super(new FunctionMetadata(
+                new Signature(
+                        "concat_ws",
+                        ImmutableList.of(),
+                        ImmutableList.of(),
+                        VarcharType.createUnboundedVarcharType().getTypeSignature(),
+                        ImmutableList.of(createUnboundedVarcharType().getTypeSignature(), createUnboundedVarcharType().getTypeSignature()),
+                        true),
+                true,
+                ImmutableList.of(new FunctionArgumentDefinition(false), new FunctionArgumentDefinition(true)),
+                false,
+                true,
+                "Concatenates elements using separator",
+                SCALAR));
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(FunctionBinding binding)
+    {
+        if (binding.getArity() < 2) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "There must be two or more arguments");
+        }
+
+        Class<?> clazz = generateConcatWsVarArgs(binding.getArity());
+        MethodHandle methodHandle = methodHandle(clazz, "concatWs", Collections.nCopies(binding.getArity(), Slice.class).toArray(new Class<?>[binding.getArity()]));
+
+        return new ScalarFunctionImplementation(
+                NULLABLE_RETURN,
+                ImmutableList.<InvocationConvention.InvocationArgumentConvention>builder()
+                        .add(NEVER_NULL)
+                        .addAll(Collections.nCopies(binding.getArity() - 1, BOXED_NULLABLE))
+                        .build(),
+                methodHandle,
+                Optional.empty());
+    }
+
+    private static Class<?> generateConcatWsVarArgs(int arity)
```

Comment:
Generating bytecode for this is overkill. This can probably be done with MethodHandle combinators.